### PR TITLE
resolve yarn.lock, bump builder version, some packaging metadata

### LINKED ISSUE #306: Updating the yarn.lock file: to not include security warnings and be generally up to date
We have a yarn.lock file that should be version controlled and updated. The yarn.lock file will be used during this step in our publishing workflow to influence the node environment that is used to build the javascript we ship (see https://github.com/jupyterhub/jupyter-server-proxy/issues/305#issuecomment-981645351).

https://github.com/jupyterhub/jupyter-server-proxy/blob/ed557ad7fc91274fe55e8db6680a30b89e4035e6/.github/workflows/publish.yaml#L19-L22

Dependabot has created a few PRs to update this file, but at this point we have the following security warnings but no dependabot PR resolving it.

image

I figured I'd try to update the versions of yarn.lock by deleting it and letting the `python setup.py sdist bdist_wheel` command recreate it for me. But, that caused an error.

```
python setup.py sdist bdist_wheel   
/home/erik/dev/contrib/jupyterhub/jupyter-server-proxy/setup.py:55: DeprecatedWarning: create_cmdclass is deprecated as of 0.8 and will be removed in 1.0. "
Use `wrap_installers` to handle prebuild steps in cmdclass.
Use `get_data_files` to handle data files.
Use `include_package_data=True` and `MANIFEST.in` for package data.

  cmdclass = create_cmdclass(
/home/erik/dev/contrib/jupyterhub/jupyter-server-proxy/setup.py:60: DeprecatedWarning: install_npm is deprecated as of 0.8 and will be removed in 1.0. Use `npm_builder` and `wrap_installers`
  install_npm(HERE / "jupyterlab-server-proxy", build_cmd="build:prod", npm=["jlpm"]),
running sdist
running jsdeps
Installing build dependencies with npm.  This may take a while...
> jlpm install
yarn install v1.21.1
info No lockfile found.
[1/4] Resolving packages...
warning @jupyterlab/application > @jupyterlab/apputils > url > querystring@0.2.0: The querystring API is considered Legacy. new code should use the URLSearchParams API instead.
warning @jupyterlab/application > @jupyterlab/ui-components > @blueprintjs/core > popper.js@1.16.1: You can find the new Popper v2 at @popperjs/core, this package is dedicated to the legacy v1
warning @jupyterlab/application > @jupyterlab/ui-components > @blueprintjs/core > react-popper > popper.js@1.16.1: You can find the new Popper v2 at @popperjs/core, this package is dedicated to the legacy v1
warning @jupyterlab/builder > @jupyterlab/buildutils > crypto@1.0.1: This package is no longer supported. It's now a built-in Node module. If you've depended on crypto, you should switch to the one that's built-in.
warning @jupyterlab/builder > @jupyterlab/buildutils > verdaccio > request@2.88.0: request has been deprecated, see https://github.com/request/request/issues/3142
warning @jupyterlab/builder > @jupyterlab/buildutils > verdaccio > @verdaccio/readme > jsdom > request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
warning @jupyterlab/builder > @jupyterlab/buildutils > verdaccio > request > har-validator@5.1.5: this library is no longer supported
warning @jupyterlab/builder > @jupyterlab/buildutils > verdaccio > @verdaccio/readme > jsdom > request > har-validator@5.1.5: this library is no longer supported
warning @jupyterlab/builder > @jupyterlab/buildutils > verdaccio > request > uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
warning @jupyterlab/builder > @jupyterlab/buildutils > verdaccio > @verdaccio/readme > jsdom > request > uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
warning @jupyterlab/builder > @jupyterlab/buildutils > verdaccio > @verdaccio/readme > jsdom > request-promise-native@1.0.9: request-promise-native has been deprecated because it extends the now deprecated request package, see https://github.com/request/request/issues/3142
[2/4] Fetching packages...
[3/4] Linking dependencies...
warning "@jupyterlab/application > @jupyterlab/ui-components@3.2.4" has unmet peer dependency "react@^17.0.1".
warning "@jupyterlab/application > @lumino/coreutils@1.11.1" has unmet peer dependency "crypto@1.0.1".
warning "@jupyterlab/application > @jupyterlab/rendermime > @jupyterlab/codemirror > y-codemirror@3.0.1" has unmet peer dependency "yjs@^13.5.17".
warning "@jupyterlab/builder > @jupyterlab/buildutils > verdaccio > clipanion@3.1.0" has unmet peer dependency "typanion@*".
warning Workspaces can only be enabled in private projects.
warning Workspaces can only be enabled in private projects.
[4/4] Building fresh packages...
⠁ 
⠁ 
error /home/erik/dev/contrib/jupyterhub/jupyter-server-proxy/jupyterlab-server-proxy/node_modules/yjs: Command failed.
Exit code: 1
Command: node ./sponsor-y.js
Arguments: 
Directory: /home/erik/dev/contrib/jupyterhub/jupyter-server-proxy/jupyterlab-server-proxy/node_modules/yjs
Output:
internal/modules/cjs/loader.js:1174
      throw new ERR_REQUIRE_ESM(filename, parentPath, packageJsonPath);
      ^

Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /home/erik/dev/contrib/jupyterhub/jupyter-server-proxy/jupyterlab-server-proxy/node_modules/yjs/sponsor-y.js
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1174:13)
    at Module.load (internal/modules/cjs/loader.js:1002:32)
    at Function.Module._load (internal/modules/cjs/loader.js:901:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:74:12)
    at internal/main/run_main_module.js:18:47 {
  code: 'ERR_REQUIRE_ESM'


Traceback (most recent call last):
  File "/home/erik/dev/contrib/jupyterhub/jupyter-server-proxy/setup.py", line 139, in <module>
    setuptools.setup(**setup_args)
  File "/home/erik/miniforge3/lib/python3.9/site-packages/setuptools/__init__.py", line 153, in setup
    return distutils.core.setup(**attrs)
  File "/home/erik/miniforge3/lib/python3.9/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/home/erik/miniforge3/lib/python3.9/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/home/erik/miniforge3/lib/python3.9/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/home/erik/miniforge3/lib/python3.9/site-packages/jupyter_packaging/setupbase.py", line 640, in run
    [self.run_command(cmd) for cmd in cmds]
  File "/home/erik/miniforge3/lib/python3.9/site-packages/jupyter_packaging/setupbase.py", line 640, in <listcomp>
    [self.run_command(cmd) for cmd in cmds]
  File "/home/erik/miniforge3/lib/python3.9/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/home/erik/miniforge3/lib/python3.9/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/home/erik/miniforge3/lib/python3.9/site-packages/jupyter_packaging/setupbase.py", line 321, in run
    c.run()
  File "/home/erik/miniforge3/lib/python3.9/site-packages/jupyter_packaging/setupbase.py", line 615, in run
    builder()
  File "/home/erik/miniforge3/lib/python3.9/site-packages/jupyter_packaging/setupbase.py", line 212, in builder
    run(npm_cmd + ['install'], cwd=node_package)
  File "/home/erik/miniforge3/lib/python3.9/site-packages/jupyter_packaging/setupbase.py", line 275, in run
    return subprocess.check_call(cmd, **kwargs)
  File "/home/erik/miniforge3/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/home/erik/miniforge3/bin/jlpm', 'install']' returned non-zero exit status 1.
```

I'd like to bump the dependencies in yarn.lock before we cut a release, so that we get rid of the security warnings. Any suggestions on how to go about this? I'm not experienced enough with the npm/yarn/jlpm tooling to know quickly what to do about this.

### PR DESCRIPTION
**Pull request by bollwyvl**
- fixes #306

Not knowing/using `jupyter-packaging` much, I did the following:

```bash
cd jupyterlab-server-proxy
rm yarn.lock
jlpm
jlpm build:prod
```

And seemed to have a working extension. I also bumped the version of `@jupyterlab/builder` to be of the line that generates `third-party-licenses.json`: as it turns out, it only ships the shims for `css-loader` and `style-loader`. As the `style/index.css` is actually empty, we _could_ ship with _no dependencies whatsoever_ without losing any functionality, which is actually quite attractive.

Apropos #298, I restored one old-school compatibility name to make `jupyter serverextension enable` work.

I had some `pip` issues (due to my environment no doubt), but rolled back any changes there except adding the relatively-new Lab trove classifiers. I've added, but commented out `JupyterLab :: 2`... if no longer tested, and nobody's crying out for it, I don't recommend we re-add it to the test matrix.

Here's hoping CI is kind!

### CODE REVIEW

**Comment by bollwyvl** [comment]
Happy to make any other changes, or add docs/explanations...

**Review by consideRatio — APPROVED** [review]
Wieee this LGTM! THANK YOU @bollwyvl !!!!!! :heart: :tada: :sunflower:

@yuvipanda @manics @ryanlovett, I figured I'd squeeze this change into 3.2.0 to get rid of a security warning.

It would feel nice to have another approval of this, because I don't understand the JS world very well, and since I'd like to include it in the 3.2.0 release.

**Comment by ryanlovett** [comment]
@consideRatio Thanks for waiting for this to merge before release! I defer to the others as well since I feel the same about JS. If there’s no activity after a long period then I can research it, though I’m guessing someone else will chime in.

**Comment by consideRatio** [comment]
Btw, I've looked into the security alerts and if they are solved by this yarn.lock update - sadly, they are not. It seems that the pinnings of @jupyterlab/application and its dependencies are too stringent and still referencing the packages with vulnerabilities.

image

Hmmm, but wait, it seems like `@jupyterlab/application` is at https://www.npmjs.com/package/@jupyterlab/application rather than 3.0.7 as we have it in this yarl.lock file.

@bollwyvl was the versions resolved to in yarl.lock in this PR recently updated? Do you understand what held back @jupyterlab/application from resolving to something more modern, like 3.2.4 than the 3.0.7 that we got? I note that there are even versions like 3.0.12 out, so it seems strange if we ended up with 3.0.7.

**Comment by bollwyvl** [comment]
I've re-pushed, replacing the `^` with `~` operator on the various ranges, and tweaked until I could get it to:
- build the typescript
- build the extension
- not complain about audits with `jlpm audit`

That being said: it's no secret that I despise the crazy version operators. I am not the first person to misunderstand them, and guarantee I won't be the last, so these might be wrong, but a look at the delta to the `yarn.lock` seems appropriate.

**Comment by manics** [comment]
https://stackoverflow.com/questions/22343224/whats-the-difference-between-tilde-and-caret-in-package-json
Shouldn't it be `^` instead of `~` to support JupyterLab `3.*`, not just `3.0.*`?

**Comment by bollwyvl** [comment]
Sure, let's try it with `^2.0 || ^3.0`.

**Comment by bollwyvl** [comment]
I generally drop stuff into an online checker, e.g. 

https://jubianchi.github.io/semver-check/#/constraint/^2.0%20||%20^3.0

**Comment by bollwyvl** [comment]
As, on closer inspection, we _are_ running the tests, I've put the trove classifier back in.

**Review by consideRatio — APPROVED** [review]
Wieeeeeeeeeeeeeeeeeeeeeee!!! Thank you so much @bollwyvl for your thorough work with this!!!!! :heart: :tada:

This LGTM - go for merge? If so, I'll go for a release of 3.2.0 including this finally - now with the security warnings resolved I think.

**Comment by maresb** [comment]
I'm really looking forward to this release!!! I wanted `rewrite_response` well over a year ago. And coincidentally just right now I have another important use case for it. So perfect timing! :smile:

**Review by bollwyvl — COMMENTED** [review]
LGTM

**Comment by consideRatio** [comment]
Aaaaahhh I'll go for a merge and release since I'm at the computer ready to do it right now and I think there is quite clear consensus on this change and to cut the release 3.2.0 with this added to it.

**Comment by ndgayan** [comment]
I'm so excited about the new release.

**Comment by consideRatio** [comment]
@ndgayan sure can, it is available on PyPI now!

**Pull request merged**
